boot/systems_test.go

```diff
@@ -0,0 +1,192 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot_test
+
+import (
+	"fmt"
+	"os"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/bootloader/bootloadertest"
+)
+
+type baseSystemsSuite struct {
+	baseBootenvSuite
+}
+
+func (s *baseSystemsSuite) SetUpTest(c *C) {
+	s.baseBootenvSuite.SetUpTest(c)
+	c.Assert(os.MkdirAll(boot.InitramfsUbuntuBootDir, 0755), IsNil)
+	c.Assert(os.MkdirAll(boot.InitramfsUbuntuSeedDir, 0755), IsNil)
+}
+
+type initramfsMarkTryRecoverySystemSuite struct {
+	baseSystemsSuite
+
+	bl *bootloadertest.MockBootloader
+}
+
+var _ = Suite(&initramfsMarkTryRecoverySystemSuite{})
+
+func (s *initramfsMarkTryRecoverySystemSuite) SetUpTest(c *C) {
+	s.baseSystemsSuite.SetUpTest(c)
+
+	s.bl = bootloadertest.Mock("bootloader", s.bootdir)
+	bootloader.Force(s.bl)
+	s.AddCleanup(func() { bootloader.Force(nil) })
+}
+
+var uncalledCheck = func() error { return fmt.Errorf("unexpected call") }
+
+func (s *initramfsMarkTryRecoverySystemSuite) testMarkRecoverySystemForRun(c *C, success bool, expectingStatus string) {
+	err := s.bl.SetBootVars(map[string]string{
+		"recovery_system_status": "try",
+		"try_recovery_system":    "1234",
+	})
+	c.Assert(err, IsNil)
+	err = boot.InitramfsMarkTryRecoverySystemResultForRunMode(success)
+	c.Assert(err, IsNil)
+
+	expectedVars := map[string]string{
+		"snapd_recovery_mode":   "run",
+		"snapd_recovery_system": "",
+
+		"recovery_system_status": expectingStatus,
+		"try_recovery_system":    "1234",
+	}
+
+	vars, err := s.bl.GetBootVars("snapd_recovery_mode", "snapd_recovery_system",
+		"recovery_system_status", "try_recovery_system")
+	c.Assert(err, IsNil)
+	c.Check(vars, DeepEquals, expectedVars)
+
+	err = s.bl.SetBootVars(map[string]string{
+		// the status it overwritten, even if it's completely bogus
+		"recovery_system_status": "foobar",
```

Comment:
if the status is really bogus (neither try or tried), should we simply clear the env and got back to run?  instead of marking as tried if it worked? it's unclear that is correct to mark is as tried if we got there without a proper status to start with

Reply:
Fair point, I've refactored InitramfsIsTryingRecoverySystem() to return a known error for a case when status is nether try or tried, and when we have a non empty status, but no reovery system.